packages/compiler-cli/src/ngtsc/typecheck/test/type_checker__completion_spec.ts

```diff
@@ -0,0 +1,138 @@
+      const MAIN_TS = absoluteFrom('/main.ts');
+      const {templateTypeChecker, programStrategy} = setup([
+        {
+          fileName: MAIN_TS,
+          templates: {'SomeCmp': `No special template needed`},
+          source: `
+            export class SomeCmp {}
+          `,
+        },
+      ]);
+      const sf = getSourceFileOrError(programStrategy.getProgram(), MAIN_TS);
+      const SomeCmp = getClass(sf, 'SomeCmp');
+
+      const [global, ...rest] =
+          templateTypeChecker.getGlobalCompletions(/* root template */ null, SomeCmp);
+      expect(rest.length).toBe(0);
+      if (global.kind !== CompletionKind.ContextComponent) {
+        return fail(`Expected a ContextComponent completion`);
+      }
+      const tcbSf =
+          getSourceFileOrError(programStrategy.getProgram(), absoluteFrom(global.shimFile));
+      const node = getTokenAtPosition(tcbSf, global.position).parent;
+      if (!ts.isExpressionStatement(node)) {
+        return fail(`Expected a ts.ExpressionStatement`);
+      }
+      expect(node.expression.getText()).toEqual('ctx.');
+      // The position should be between the '.' and a following space.
+      expect(tcbSf.text.substr(global.position - 1, 2)).toEqual('. ');
+    });
+
+    it('should return additional completions for references and variables when available', () => {
+      const MAIN_TS = absoluteFrom('/main.ts');
+      const {templateTypeChecker, programStrategy} = setup([
+        {
+          fileName: MAIN_TS,
+          templates: {
+            'SomeCmp': `
+              <div *ngFor="let user of users">
+                <div #innerRef></div>
+                <div *ngIf="user">
+                  <div #notInScope></div>
+                </div>
+              </div>
+              <div #topLevelRef></div>
+          `
+          },
+          source: `
+            export class SomeCmp {
+              users: string[];
+            }
+          `,
+        },
+      ]);
+      const sf = getSourceFileOrError(programStrategy.getProgram(), MAIN_TS);
+      const SomeCmp = getClass(sf, 'SomeCmp');
+
+      const tmpl = templateTypeChecker.getTemplate(SomeCmp)!;
+      const ngForTemplate = tmpl[0] as TmplAstTemplate;
+
+      const [contextCmp, ...rest] =
+          templateTypeChecker.getGlobalCompletions(ngForTemplate, SomeCmp);
+      if (contextCmp.kind !== CompletionKind.ContextComponent) {
+        return fail(`Expected first completion to be a ContextComponent`);
+      }
+
+      const completionKeys: string[] = [];
+      for (const completion of rest) {
+        if (completion.kind !== CompletionKind.Reference &&
+            completion.kind !== CompletionKind.Variable) {
+          return fail(`Unexpected CompletionKind, expected a Reference or Variable`);
+        }
+        completionKeys.push(completion.node.name);
+      }
+
+      expect(completionKeys).toEqual(['innerRef', 'user', 'topLevelRef']);
```

Comment:
```suggestion
      expect(completionKeys.sort()).toEqual(['innerRef', 'user', 'topLevelRef'].sort());
```
Unless the order matters here, consider using `sort`. The order might actually matter for prioritizing completions in the language service though...